### PR TITLE
HMAC variants of sha256/sha512 implemented

### LINKED ISSUE #7297: Implement HMAC variants of std/hash algorithms
The WebAssembly  "createHash" implementation of std/hash is missing HMAC variants.
Should be able to provide this quite easily by taking in the hmac crate and exposing hmac<sha256> etc.

### PR DESCRIPTION
**Pull request by mustafapc19**
Fixes #7297

### CODE REVIEW

**Comment by mustafapc19** [comment]
@caspervonb please forgive all the lame spam commits :)

**Comment by caspervonb** [comment]
> @caspervonb please forgive all the lame spam commits :)

No worries, everything gets squashed on merge in here regardless 🙃 
I'll review this later today, thanks!

**Comment by caspervonb** [comment]
Could you merge this with master? We had a CI glitch and this is still stuck on that.

**Comment by mustafapc19** [comment]
> Could you merge this with master? We had a CI glitch and this is still stuck on that.

@caspervonb Was this directed to me? I don't have the authority to merge R8?

**Comment by caspervonb** [comment]
> @caspervonb Was this directed to me? I don't have the authority to merge R8? Or do suggest to pull latest changes and send pr again?

It was, I meant merge denoland's master into your branch as the CI is broken on the revision you forked from 👍

**Comment by kitsonk** [comment]
>  I meant merge denoland's master into your branch

Or a re-base... either way.

File: std/hash/_wasm/hmac_hash.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import init, { HmacSha256Hash, HmacSha512Hash, source } from "./wasm.js";
+
+import type { HmacHasher, Message } from "../hmac_hasher.ts";
+
+await init(source);
+
+type HmacHashMethods = HmacSha256Hash | HmacSha512Hash;
+
+export class HmacHash implements HmacHasher {
```
Comment (caspervonb) [inline std/hash/_wasm/hmac_hash.ts:11]: Already generating a wrapper class with wasm_bindgen, wrapping the wrapper class seems like one level of indirection too many.

File: std/hash/_wasm/src/lib.rs [hunk 2]
```diff
@@ -49,3 +51,71 @@ pub fn update_hash(hash: &mut DenoHash, data: &[u8]) {
 pub fn digest_hash(hash: &mut DenoHash) -> Box<[u8]> {
   hash.inner.finalize_reset()
 }
+
+#[wasm_bindgen]
+pub struct HmacSha256Hash {
```
Comment (caspervonb) [inline std/hash/_wasm/src/lib.rs:56]:
```suggestion
pub struct HmacSha256 {
```

File: std/hash/_wasm/src/lib.rs [hunk 3]
```diff
@@ -49,3 +51,71 @@ pub fn update_hash(hash: &mut DenoHash, data: &[u8]) {
...
+  inner: Hmac<Sha256>,
+}
+
+#[wasm_bindgen]
+impl HmacSha256Hash {
```
Comment (caspervonb) [inline std/hash/_wasm/src/lib.rs:61]:
```suggestion
impl HmacSha256 {
```

File: std/hash/_wasm/src/lib.rs [hunk 4]
```diff
@@ -49,3 +51,71 @@ pub fn update_hash(hash: &mut DenoHash, data: &[u8]) {
...
+      .finalize_reset()
+      .into_bytes()
+      .as_slice()
+      .to_vec()
+      .into_boxed_slice()
+  }
+}
+
+#[wasm_bindgen]
+pub struct HmacSha512Hash {
```
Comment (caspervonb) [inline std/hash/_wasm/src/lib.rs:90]:
```suggestion
pub struct HmacSha512 {
```

File: std/hash/_wasm/src/lib.rs [hunk 5]
```diff
@@ -49,3 +51,71 @@ pub fn update_hash(hash: &mut DenoHash, data: &[u8]) {
...
+  inner: Hmac<Sha512>,
+}
+
+#[wasm_bindgen]
+impl HmacSha512Hash {
```
Comment (caspervonb) [inline std/hash/_wasm/src/lib.rs:95]:
```suggestion
impl HmacSha512 {
```

**Review by caspervonb — COMMENTED** [review]
Some minor nits, I can clean these up if you can't find the time.

**Comment by mustafapc19** [comment]
@caspervonb I got really busy with college and stuff. What is happening? I am confused. What should I do?

**Pull request closed without merging**